Apply growth rates to initial_value datasets

daniel-thom opened the pull request:
## Description
This PR removes the ability to perform per-dataset aggregations
prior to combining datasets. That methodology did not work with growth
rate datasets. Secondly, while there are some efficiencies to be gained
with aggregating dimensions in that early stage, such as combining
electricity end uses, we expect that we will usually store cached
versions of project-mapped datasets. User queries will use those cached
datasets, and so this is not a significant loss in functionality.

The PR also fixes two issues in the current codebase that caused #180 and #181:
- Fraction values were not being applied to datasets after dimension mapping. This resulted in duplicate combinations of dimension rows. This PR applies the fraction with groupBy/agg.
- If a query filtered a dimension with specific records but a dataset includes dimension mappings that specify records in the to_id column that should be filtered out, those invalid records would show up in the final table. This PR fixes that by filtering the dimension mapping records.

This PR includes #182. That PR should be reviewed and merged first.

This is currently draft because of open questions regarding the data_source dimension.

daniel-thom commented on dsgrid/dataset/growth_rates.py:
```diff
@@ -0,0 +1,92 @@
+logger = logging.getLogger(__name__)
+
+
+def apply_growth_rate_123(
```
@lixiliu The names here are obviously bogus. How would you like to deal with them? Would you like to takeover and create your own structure of name/function mappings?

lixiliu replied:
here or in new PR?

elainethale replied:
For now we could call this one apply_annual_growth_rate?

daniel-thom replied:
If @lixiliu is going to implement this soon, I would prefer to do nothing.

daniel-thom commented on dsgrid/query/models.py:
```diff
@@ -213,40 +205,134 @@ def handle_dimension_filters(cls, dimension_filters):
+    def get_dataset_id(self) -> str:
+        return self.dataset_id
+
+    def iter_dataset_ids(self):
```
Stale and needs deletion

daniel-thom commented on dsgrid/query/models.py:
```diff
@@ -269,16 +355,17 @@ def check_unsupported_fields(cls, values):
-        dataset_ids = values.get("dataset_ids")
-        if dataset_ids is None:
-            return excluded_dataset_ids
+    # TODO: currently incorrect and may get deleted
```
handle

daniel-thom commented on tests/test_datasets.py:
```diff
@@ -25,24 +24,23 @@
 
 @pytest.fixture(scope="module")
-def setup_registry(make_test_project_dir_module, make_test_data_dir_module):
+def setup_registry(tmp_path_factory, make_test_project_dir_module, make_test_data_dir_module):
```
This is an unrelated improvement. It uses a pytest fixture instead of making our own TemporaryDirectory.

daniel-thom commented on tests/test_queries.py:
```diff
@@ -650,41 +672,43 @@ def make_query(self):
+        assert math.isclose(val, expected_values["la_electricity_hour_16"])
 
 
-class QueryTestDiurnalElectricityUseByCountyPrePostConcat(QueryTestBase):
```
This test case is no longer valid because it relied on per-dataset aggregations.

daniel-thom commented on dsgrid/dataset/growth_rates.py:
```diff
@@ -0,0 +1,92 @@
+        ).drop(column)
+
+    dim_columns = set(initial_value_df.columns) - pivoted_columns - time_columns
+    # TODO DT: data_source needs some thought. They are different in these two dfs.
```
Per team discussion today, we will leave this as-is. Meghan and Lixi want this data_source value to change to either `derived` or a concatenation of the two data sources. That presents some complications for registering the resulting derived dataset. We will leave that open for now. #185

daniel-thom commented on dsgrid/dataset/growth_rates.py:
```diff
@@ -0,0 +1,92 @@
+            f"ExponentialGrowthDatasetModel base_year={dataset.base_year} is not in {iv_years_sorted}"
+        )
+
+    if len(iv_years) > 1:
+        # TODO: is this logic correct?
+        # TODO: needs test case
+        initial_value_df = initial_value_df.filter(f"{model_year_column} == '{base_year}'")
```
@lixiliu Please confirm this.

lixiliu replied:
looks good to me

daniel-thom commented on dsgrid/utils/dataset.py:
```diff
@@ -16,21 +16,19 @@
 def map_and_reduce_stacked_dimension(df, records, column):
     if "fraction" not in df.columns:
-        df = df.withColumn("fraction", F.lit(1))
+        df = df.withColumn("fraction", F.lit(1.0))
```
I changed this because it felt weird to add an integer column that should be a float. It was working correctly, but I think this reads better.

daniel-thom commented on dsgrid/utils/dataset.py:
```diff
@@ -96,9 +98,11 @@ def map_and_reduce_pivoted_dimension(df, records, pivoted_columns, operation, re
-            expr = f"{val} AS {column}"
+        # TODO: Need to decide how to handle NULL values. If they should not be included in the
+        # mean, the logic below is incorred.t
+        # elif operation in ("avg", "mean"):
```
@lixiliu Can we remove this section? I'm not imagining that anyone would find this useful. (finding the mean of metric values in one row, for example)

lixiliu replied:
Isn't this mean of pivoted columns?

daniel-thom replied:
If the pivoted columns are metric, such as electricity_heating and electricity_pumps, would anyone want the average of those two fields? Then suppose transportation is included. Would the mean be calculated only on the non-null columns? We would need to write SQL logic to handle all of that.

lixiliu replied:
For completeness, I think we do, doesn't mean we have to address in this PR.

lixiliu commented on dsgrid/dataset/growth_rates.py:
```diff
@@ -0,0 +1,92 @@
+    if DimensionType.DATA_SOURCE.value in gr_df.columns:
+        gr_df = gr_df.drop(DimensionType.DATA_SOURCE.value)
+
+    df = initial_value_df.join(gr_df, on=list(dim_columns))
```
I think you (or I) could use the new formula functionality here

daniel-thom replied:
Are you talking about DatasetExpressionHandler?

lixiliu replied:
The multiplication of value col and _gr col

lixiliu replied:
Yes, this PR doesn't have to address this

lixiliu commented on dsgrid/project.py:
```diff
@@ -183,32 +170,26 @@ def process_query(self, context: QueryContext, cached_project_mapped_datasets_di
-                    )
-                main_df = main_df.union(df)
+            datasets[dataset_id] = DatasetExpressionHandler(
+                df.select(*final_columns), time_columns + dim_columns, pivoted_columns_sorted
```
I think you can just pass in df into DatasetExpressionHandler here, ordering needs to be enforced at the final output.

daniel-thom replied:
When you perform DataFrame.union the dataframes have to have identical column order. This ensure that happens. Or am I missing something?

lixiliu replied:
I think this just turns it into an object. The ordering here doesn't matter I think. This comment goes with the next comment

daniel-thom replied:
Is the issue that I'm putting dimension columns after pivoted columns in project.py but then before pivoted columns in the expression handler? That can be fixed. The main thing is that the columns in all datasets have to be in the same order when you perform the union, otherwise you get garbage.

daniel-thom replied:
I fixed column ordering in the dataset expression code - that shouldn't change order. I'm very tempted to stop worrying about order in our processing code until the end, however. Every time you join, Spark puts the join columns first.

lixiliu replied:
Yes, I agree, for union, we could just have a utility func that attempts to make the cols in two df the same, and the final ordering can be a last step.

lixiliu commented on dsgrid/project.py:
```diff
@@ -183,32 +170,26 @@ def process_query(self, context: QueryContext, cached_project_mapped_datasets_di
+            )
 
-        return main_df
+        return evaluate_expression(context.model.project.dataset.expression, datasets).df
```
I think we need to do `df.select(*final_columns)` here after `evaluate_expression()`

daniel-thom replied:
They have to be in the correct order before evaluate_expression; otherwise the result is garbage.

lixiliu replied:
I think this is a stale comment now since `final_columns` is not defined and not being enforced here.

lixiliu commented on dsgrid/project.py:
```diff
@@ -220,9 +201,151 @@ def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+            model_file = hash_dir / "model.json"
+            model_file.write_text(text)
+        cached_dataset_path = hash_dir / (dataset_id + ".parquet")
+        metadata_file = Path(str(cached_dataset_path).replace(".parquet", ".json5"))
```
`cached_dataset_path.with_suffix(".json5")`: https://docs.python.org/3/library/pathlib.html

daniel-thom replied:
Yeah, thanks. I just used that one two weeks ago elsewhere.

lixiliu commented on dsgrid/project.py:
```diff
@@ -220,9 +201,151 @@ def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+        logger.info("Finished processing query for dataset_id=%s", dataset_id)
+        return cached_dataset_path
+
+    def _process_exponential_growth_dataset(
```
Is this temp or are we expecting to create one of these for each new dataset construction type?

daniel-thom replied:
I'm expecting that we will need a custom function for any type that cannot be directly handled with DatasetExpressionHandler.

lixiliu commented on dsgrid/project.py:
```diff
@@ -220,9 +201,151 @@ def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+            )
+
+        model_year_column = get_myear_column(dataset.initial_value_dataset_id)
+        model_year_column_gr = get_myear_column(dataset.growth_rate_dataset_id)
```
How would this only have 1 single model_year?

daniel-thom replied:
This is the model year column, which is a dimension query name. Or am I missing your question?

lixiliu commented on dsgrid/query/models.py:
```diff
@@ -213,40 +205,119 @@ def handle_dimension_filters(cls, dimension_filters):
+    dataset_id: str = Field(description="Dataset identifier")
+    dataset_type: DatasetType = Field(
+        description="Type of dataset specified in a query",
+        default=DatasetType.STANDALONE,
```
Specifying the default value here feels weird to me. But maybe it's just how it's done

daniel-thom replied:
Got a better suggestion?
1. The value `standalone` needs to be in the JSON so that we know how to deserialize.
2. We don't want to make the Python user type `StandaloneDatasetModel(dataset_type=DatasetType.STANDALONE)`

lixiliu replied:
No

lixiliu commented on dsgrid/utils/dataset.py:
```diff
@@ -96,9 +98,11 @@ def map_and_reduce_pivoted_dimension(df, records, pivoted_columns, operation, re
-        elif operation in ("avg", "mean"):
-            val = "(" + "+".join(expr) + f") / {len(expr)}"
-            expr = f"{val} AS {column}"
+        # TODO: Need to decide how to handle NULL values. If they should not be included in the
```
Yep

daniel-thom replied:
Is this an answer to my question below?

lixiliu commented on tests/test_dataset_utils.py:
```diff
@@ -0,0 +1,132 @@
+    assert res.filter("county == 'Adams'").collect()[0].all_electricity == 4.8
+
+
+def test_map_and_reduce_pivoted_dimension_min(dataframes):
```
Do we have one for map_and_reduce_stacked_dimensions?

daniel-thom replied:
We do not. We should add something similar.

lixiliu replied:
Add TODO?

lixiliu reviewed (commented):
🥇 quality stuff

daniel-thom commented on dsgrid/cli/query.py:
```diff
@@ -179,7 +184,13 @@ def create_project(
+                dataset_id=dataset_id,
+                source_datasets=[
+                    StandaloneDatasetModel(dataset_type=DatasetType.STANDALONE, dataset_id=x)
+                    for x in project.config.list_registered_dataset_ids()
```
This is creating a template query for the user. There is probably a lot more we can customize. I'm not too concerned now.

daniel-thom commented on dsgrid/dataset/dataset_schema_handler_base.py:
```diff
@@ -246,8 +257,51 @@ def _get_dataset_to_project_mapping_reference(self, dimension_type: DimensionTyp
+
+    def _remap_dimension_columns(self, df, pivoted_columns=None, filtered_records=None):
         if pivoted_columns is None:
             pivoted_columns = set(self.get_pivoted_dimension_columns())
```
It's weird that this function can get called on load_data_lookup when there are no pivoted columns. If we make a change, I think we should separate the functionality. If pivoted columns are needed, it should be provided by the caller.

daniel-thom replied:
I changed my mind and agree with you.

daniel-thom commented on dsgrid/dataset/dataset_schema_handler_base.py:
```diff
@@ -289,6 +348,21 @@ def _remap_dimension_columns(self, df, pivoted_columns=None):
+            )
+        # Maintain column order.
+        agg_ops = [agg_func(x).alias(x) for x in [y for y in df.columns if y in pivoted_columns]]
+        gcols = set(df.columns) - pivoted_columns - {"fraction"}
```
Maybe you wrote this before my last commit. I think it is now correct.

daniel-thom commented on dsgrid/dataset/dataset_schema_handler_one_table.py:
```diff
@@ -145,18 +147,52 @@ def make_project_dataframe(self, project_config):
+            ld_df, pivoted_columns=pivoted_columns, filtered_records=context.get_record_ids()
+        )
+
+        pivoted_columns = set(ld_df.columns).intersection(
```
I agree that it could be put inside _remap, but we should have different remaps for with-pivoted and without-pivoted.

daniel-thom commented on dsgrid/dataset/growth_rates.py:
```diff
@@ -0,0 +1,92 @@
+            F.pow((1 + F.col(column)), F.col(model_year_column).cast(IntegerType()) - base_year),
+        ).drop(column)
+
+    dim_columns = set(initial_value_df.columns) - pivoted_columns - time_columns
```
After the commit I just pushed, this should be correct. Let me know if you disagree.

lixiliu replied:
This is just a join key, so not relevant

elainethale commented on dsgrid/project.py:
```diff
@@ -220,9 +201,151 @@ def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+            if dataset.construction_method == "formula123":
+                df = apply_growth_rate_123(
+                    dataset, iv_df, gr_df, time_columns, model_year_column, pivoted_columns
+                )
```
Ick. Seems like the string tag should match the function name? "apply_growth_rate_123" not "formula123"? And there should at least be a dict somewhere there that provides the look up?

daniel-thom replied:
This is a placeholder. @lixiliu agreed to create a naming convention in a subsequent PR.

elainethale replied:
okay

daniel-thom replied:
Perhaps I should not have deleted dsgrid_project/datasets/modeled/resstock/dimension_mappings/model_year_to_model_year.csv in https://github.com/dsgrid/dsgrid-project-StandardScenarios/pull/71.

elainethale reviewed (changes requested):
`git submodule update` maybe doesn't work on Windows:

```
C:\projects\dsgrid>git submodule update
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimension_mappings/aeo2021-commercial-building-types__conus-2022-detailed-subsectors__cdf5cec9-8ac1-4d54-89aa-36c5c38a7be3/1.0.0/lookup_comstock_to_aeo2021_com_bldg_types.csv: Filename too long
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimension_mappings/aeo2021-commercial-energy-enduses-unitless__conus-2022-detailed-end-uses-kwh__8b903847-45c5-4ac5-9d85-1b0a3200391c/1.0.0/lookup_comstock_to_aeo2021_com_enduses.csv: Filename too long
warning: unable to access 'filtered_registries/simple_standard_scenarios/configs/dimension_mappings/aeo2021-residential-energy-enduses-unitless_for-growth-rates__conus-2022-detailed-end-uses-kwh__22f3f031-822f-4952-8fc7-7b7c1fd77b3e/1.0.0/.gitattributes': Filename too long
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimension_mappings/aeo2021-residential-energy-enduses-unitless_for-growth-rates__conus-2022-detailed-end-uses-kwh__22f3f031-822f-4952-8fc7-7b7c1fd77b3e/1.0.0/dimension_mapping.json5: Filename too long
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimension_mappings/aeo2021-residential-energy-enduses-unitless_for-growth-rates__conus-2022-detailed-end-uses-kwh__22f3f031-822f-4952-8fc7-7b7c1fd77b3e/1.0.0/lookup_aeo2021_res_combined_to_resstock_enduses.csv: Filename too long
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimension_mappings/all_dsgrid_conus_2022_geographies__us_counties_2020_l48__3d38ac19-9a9d-4b2d-8b5d-5bce98ca7fb8/1.0.0/reverse_us_counties_2020_l48__all_dsgrid_conus_2022_geographies__dd3cd322-a61e-453d-8324-754a3b8211d3.csv: Filename too long
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimension_mappings/data-sources_for_dsgrid_standard_scenarios_2021__all_dsgrid_conus_2022_data_sources__6fa794e8-7b35-496e-9bc1-921497000c15/1.0.0/dimension_mapping.json5: Filename too long
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimension_mappings/data-sources_for_dsgrid_standard_scenarios_2021__all_dsgrid_conus_2022_data_sources__6fa794e8-7b35-496e-9bc1-921497000c15/1.0.0/lookup_data_source_to_all_data_sources.csv: Filename too long
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimension_mappings/us_census_divisions__us_counties_2020_l48__1b7d07dc-2ca5-4d02-9597-de535549a7cc/1.0.0/reverse_us_counties_2020_l48__us_census_divisions__a26bc660-c5a6-4e64-93ff-df3039d39732.csv: Filename too long
error: unable to create file filtered_registries/simple_standard_scenarios/configs/dimensions/metric/aeo2021-residential-energy-enduses-unitless_for-growth-rates__09d27acb-3076-4a73-9ba1-93a44e405dd8/1.0.0/aeo2021_residential_condensed_energy_enduses.csv: Filename too long
Submodule path 'dsgrid-test-data': checked out 'b91b2e11d66afd33289273a8bab36f278ff4f372'
```

And maybe model_year filtering doesn't work in conjunction with ExponentialGrowthDatasetModel? Is that expected? This query:

```json
{
  "name": "Taos County Projections",
  "version": "0.1.0",
  "project": {
    "project_id": "dsgrid_conus_2022",
    "dataset": {
      "dataset_id": "taos_county_select_years",
      "source_datasets": [
        {
          "dataset_id": "tempo_conus_2022",
          "dataset_type": "standalone"
        },
        {
          "dataset_id": "comstock_projected_conus_2022",
          "initial_value_dataset_id": "comstock_conus_2022_reference",
          "growth_rate_dataset_id": "aeo2021_reference_commercial_energy_use_growth_factors",
          "construction_method": "formula123",
          "base_year": null,
          "dataset_type": "exponential_growth"
        },
        {
          "dataset_id": "resstock_projected_conus_2022",
          "initial_value_dataset_id": "resstock_conus_2022_reference",
          "growth_rate_dataset_id": "aeo2021_reference_residential_energy_use_growth_factors",
          "construction_method": "formula123",
          "base_year": null,
          "dataset_type": "exponential_growth"
        }
      ],
      "expression": "tempo_conus_2022 | comstock_projected_conus_2022 | resstock_projected_conus_2022",
      "params": {
        "dimension_filters": [
          {
            "dimension_type": "geography",
            "dimension_query_name": "county",
            "column": "id",
            "operator": "isin",
            "value": [
              "35055"
            ],
            "negate": false,
            "filter_type": "DimensionFilterColumnOperatorModel"
          },
          {
            "dimension_type": "model_year",
            "dimension_query_name": "model_year",
            "column": "id",
            "operator": "isin",
            "value": [
              "2020",
              "2025",
              "2030",
              "2035",
              "2040",
              "2045",
              "2050"
            ],
            "negate": false,
            "filter_type": "DimensionFilterColumnOperatorModel"
          }
        ],
        "table_format": "pivoted"
      }
    },
    "excluded_dataset_ids": [],
    "include_dsgrid_dataset_components": false,
    "version": null,
    "spark_conf_per_dataset": []
  },
  "result": {
    "supplemental_columns": [],
    "replace_ids_with_names": false,
    "aggregations": [
      {
        "aggregation_function": "sum",
        "dimensions": {
          "data_source": [],
          "geography": [
            {
              "dimension_query_name": "county",
              "function": null,
              "alias": null
            }
          ],
          "metric": [
            {
              "dimension_query_name": "electricity",
              "function": null,
              "alias": null
            }
          ],
          "model_year": [
            {
              "dimension_query_name": "model_year",
              "function": null,
              "alias": null
            }
          ],
          "scenario": [
            {
              "dimension_query_name": "scenario",
              "function": null,
              "alias": null
            }
          ],
          "sector": [
            {
              "dimension_query_name": "sector",
              "function": null,
              "alias": null
            }
          ],
          "subsector": [
            {
              "dimension_query_name": "all_subsectors",
              "function": null,
              "alias": null
            }
          ],
          "time": [
            {
              "dimension_query_name": "time_est",
              "function": null,
              "alias": null
            }
          ],
          "weather_year": [
            {
              "dimension_query_name": "weather_2012",
              "function": null,
              "alias": null
            }
          ]
        }
      }
    ],
    "reports": [],
    "output_format": "parquet",
    "sort_columns": [],
    "dimension_filters": [],
    "time_zone": null
  }
}
```

Yields error:

```bash
23/01/11 08:49:30 INFO StandaloneSchedulerBackend: SchedulerBackend is ready for scheduling beginning after reached minRegisteredResourcesRatio: 0.0
Traceback (most recent call last):
  File "/home/ehale/c-leap/dsgrid_taos_county_query.py", line 158, in <module>
    main(args.output_dir)
  File "/home/ehale/c-leap/dsgrid_taos_county_query.py", line 148, in main
    ProjectQuerySubmitter(project, output_dir).submit(
  File "/home/ehale/dsgrid/dsgrid/utils/timing.py", line 130, in timed_
    return _timed_func(collector, func, *args, **kwargs)
  File "/home/ehale/dsgrid/dsgrid/utils/timing.py", line 139, in _timed_func
    return func(*args, **kwargs)
  File "/home/ehale/dsgrid/dsgrid/query/query_submitter.py", line 255, in submit
    return self._run_query(
  File "/home/ehale/dsgrid/dsgrid/query/query_submitter.py", line 135, in _run_query
    df = self._project.process_query(context, self._cached_project_mapped_datasets_dir())
  File "/home/ehale/dsgrid/dsgrid/utils/timing.py", line 130, in timed_
    return _timed_func(collector, func, *args, **kwargs)
  File "/home/ehale/dsgrid/dsgrid/utils/timing.py", line 139, in _timed_func
    return func(*args, **kwargs)
  File "/home/ehale/dsgrid/dsgrid/project.py", line 146, in process_query
    path = self._process_exponential_growth_dataset(
  File "/home/ehale/dsgrid/dsgrid/project.py", line 279, in _process_exponential_growth_dataset
    self._build_exponential_growth_dataset(
  File "/home/ehale/dsgrid/dsgrid/utils/timing.py", line 130, in timed_
    return _timed_func(collector, func, *args, **kwargs)
  File "/home/ehale/dsgrid/dsgrid/utils/timing.py", line 139, in _timed_func
    return func(*args, **kwargs)
  File "/home/ehale/dsgrid/dsgrid/project.py", line 336, in _build_exponential_growth_dataset
    df = apply_growth_rate_123(
  File "/home/ehale/dsgrid/dsgrid/dataset/growth_rates.py", line 39, in apply_growth_rate_123
    initial_value_df, base_year = _check_model_years(
  File "/home/ehale/dsgrid/dsgrid/dataset/growth_rates.py", line 77, in _check_model_years
    base_year = iv_years_sorted[0]
IndexError: list index out of range
```

On the plus side, I can confirm that #180 and #181 are resolved.

daniel-thom commented:
Is the "filename too long" issue the same as https://stackoverflow.com/questions/22575662/filename-too-long-in-git-for-windows?

elainethale commented:
> Is the "filename too long" issue the same as https://stackoverflow.com/questions/22575662/filename-too-long-in-git-for-windows?

Probably, and now I can't replicate the issue (to even confirm that the git config command fixes it), so let's assume that's not an issue.

daniel-thom commented on dsgrid/dataset/dataset_expression_handler.py:
```diff
@@ -23,10 +23,19 @@ def _op(self, other, op):
-            *self.dimension_columns, *expr
-        )
+        def renamed(col):
+            return col + "_other"
```
The rest of the variables in this function use `other`. And the columns get dropped.

daniel-thom commented:
@elainethale @lixiliu The model_year issue above happens because we removed the mappings in dsgrid_project/datasets/modeled/resstock/dimension_mappings/model_year_to_model_year.csv in this PR: https://github.com/dsgrid/dsgrid-project-StandardScenarios/pull/71/.

Either we restore those changes or figure out a different way to filter datasets. We don't need that mapping for any case other than this, I think. We removed it because we didn't think it was necessary. Any other thoughts?

elainethale commented on dsgrid/dataset/dataset_schema_handler_base.py:
```diff
@@ -246,17 +258,64 @@ def _get_dataset_to_project_mapping_reference(self, dimension_type: DimensionTyp
+        return df
+
+    def _prefilter_time_dimension(self, context: QueryContext, df):
+        # TODO:
```
Sounds good

elainethale commented:
> @elainethale @lixiliu The model_year issue above happens because we removed the mappings in dsgrid_project/datasets/modeled/resstock/dimension_mappings/model_year_to_model_year.csv in this PR: [dsgrid/dsgrid-project-StandardScenarios#71](https://github.com/dsgrid/dsgrid-project-StandardScenarios/pull/71).
> 
> Either we restore those changes or figure out a different way to filter datasets. We don't need that mapping for any case other than this, I think. We removed it because we didn't think it was necessary. Any other thoughts?

Is filtering happening before or after applying the growth rate formula? It seems like filtering should work if it happens after the growth rate formula is applied?

daniel-thom commented:
> > @elainethale @lixiliu The model_year issue above happens because we removed the mappings in dsgrid_project/datasets/modeled/resstock/dimension_mappings/model_year_to_model_year.csv in this PR: [dsgrid/dsgrid-project-StandardScenarios#71](https://github.com/dsgrid/dsgrid-project-StandardScenarios/pull/71).
> > Either we restore those changes or figure out a different way to filter datasets. We don't need that mapping for any case other than this, I think. We removed it because we didn't think it was necessary. Any other thoughts?
> 
> Is filtering happening before or after applying the growth rate formula? It seems like filtering should work if it happens after the growth rate formula is applied?

We fully process the standalone dataset and growth rate set, including filter application, before applying the growth rate. I think that is what we want. If the user cares about a single county, as in this case, it's better to pre-filter the dataset because it saves a lot of work.

If we restore the mapping then the dataset itself will get exploded in a pointless way (40+ years of redundant data). This needs some more thought.

elainethale commented:
I was hoping that this query:

```json
{
  "name": "Taos County Projections",
  "version": "0.1.0",
  "project": {
    "project_id": "dsgrid_conus_2022",
    "dataset": {
      "dataset_id": "taos_county_select_years",
      "source_datasets": [
        {
          "dataset_id": "tempo_conus_2022",
          "dataset_type": "standalone"
        },
        {
          "dataset_id": "comstock_projected_conus_2022",
          "initial_value_dataset_id": "comstock_conus_2022_reference",
          "growth_rate_dataset_id": "aeo2021_reference_commercial_energy_use_growth_factors",
          "construction_method": "formula123",
          "base_year": null,
          "dataset_type": "exponential_growth"
        },
        {
          "dataset_id": "resstock_projected_conus_2022",
          "initial_value_dataset_id": "resstock_conus_2022_reference",
          "growth_rate_dataset_id": "aeo2021_reference_residential_energy_use_growth_factors",
          "construction_method": "formula123",
          "base_year": null,
          "dataset_type": "exponential_growth"
        }
      ],
      "expression": "tempo_conus_2022 | comstock_projected_conus_2022 | resstock_projected_conus_2022",
      "params": {
        "dimension_filters": [
          {
            "dimension_type": "geography",
            "dimension_query_name": "county",
            "column": "id",
            "operator": "isin",
            "value": [
              "35055"
            ],
            "negate": false,
            "filter_type": "DimensionFilterColumnOperatorModel"
          }
        ],
        "table_format": "pivoted"
      }
    },
    "excluded_dataset_ids": [],
    "include_dsgrid_dataset_components": false,
    "version": null,
    "spark_conf_per_dataset": []
  },
  "result": {
    "supplemental_columns": [],
    "replace_ids_with_names": false,
    "aggregations": [
      {
        "aggregation_function": "sum",
        "dimensions": {
          "data_source": [],
          "geography": [
            {
              "dimension_query_name": "county",
              "function": null,
              "alias": null
            }
          ],
          "metric": [
            {
              "dimension_query_name": "electricity",
              "function": null,
              "alias": null
            }
          ],
          "model_year": [
            {
              "dimension_query_name": "model_year",
              "function": null,
              "alias": null
            }
          ],
          "scenario": [
            {
              "dimension_query_name": "scenario",
              "function": null,
              "alias": null
            }
          ],
          "sector": [
            {
              "dimension_query_name": "sector",
              "function": null,
              "alias": null
            }
          ],
          "subsector": [
            {
              "dimension_query_name": "all_subsectors",
              "function": null,
              "alias": null
            }
          ],
          "time": [
            {
              "dimension_query_name": "time_est",
              "function": null,
              "alias": null
            }
          ],
          "weather_year": [
            {
              "dimension_query_name": "weather_2012",
              "function": null,
              "alias": null
            }
          ]
        }
      }
    ],
    "reports": [],
    "output_format": "parquet",
    "sort_columns": [],
    "dimension_filters": [
      {
        "dimension_type": "model_year",
        "dimension_query_name": "model_year",
        "column": "id",
        "operator": "isin",
        "value": [
          "2020",
          "2025",
          "2030",
          "2035",
          "2040",
          "2045",
          "2050"
        ],
        "negate": false,
        "filter_type": "DimensionFilterColumnOperatorModel"
      }
    ],
    "time_zone": null
  }
}
```

would be a workaround to my filtering issue, but I got this error:

```bash
23/01/11 13:43:28 INFO DAGScheduler: Job 9 finished: collect at /home/ehale/dsgrid/dsgrid/utils/spark.py:272, took 0.094841 s
Traceback (most recent call last):
  File "/home/ehale/c-leap/dsgrid_taos_county_query.py", line 160, in <module>
    main(args.output_dir)
  File "/home/ehale/c-leap/dsgrid_taos_county_query.py", line 150, in main
    ProjectQuerySubmitter(project, output_dir).submit(
  File "/home/ehale/dsgrid/dsgrid/utils/timing.py", line 130, in timed_
    return _timed_func(collector, func, *args, **kwargs)
  File "/home/ehale/dsgrid/dsgrid/utils/timing.py", line 139, in _timed_func
    return func(*args, **kwargs)
  File "/home/ehale/dsgrid/dsgrid/query/query_submitter.py", line 255, in submit
    return self._run_query(
  File "/home/ehale/dsgrid/dsgrid/query/query_submitter.py", line 155, in _run_query
    df = self._apply_filters(df, context)
  File "/home/ehale/dsgrid/dsgrid/query/query_submitter.py", line 176, in _apply_filters
    raise Exception("Filtering result data is not supported yet")
Exception: Filtering result data is not supported yet
```

elainethale reviewed (approved):
I'm okay moving forward as long as we capture issues that need fixing in subsequent PRs.

lixiliu commented on dsgrid/project.py:
```diff
@@ -183,32 +170,26 @@ def process_query(self, context: QueryContext, cached_project_mapped_datasets_di
         dim_columns = sorted(dim_columns)
         time_columns = sorted(time_columns)
         expected_columns = time_columns + pivoted_columns_sorted + dim_columns
```
NIT: I think expected_columns should be `time_columns + dim_columns + pivoted_columns_sorted` to match ordering in DatasetExpressionHandler. I think this func here doesn't enforce final cols ordering anymore, it just makes things less confusing since DataExpressionHandler would otherwise change col ordering.

lixiliu reviewed (approved):
LGTM, thanks!

daniel-thom commented on dsgrid/cli/query.py:
```diff
@@ -257,16 +261,15 @@ def validate_project(query_file):
 @click.option(
-    "-p",
-    "--persist-intermediate-table",
+    "--persist-intermediate-table/--no-persist-intermediate-table",
```
Yes, you can do that. But which of these help options do you prefer ():
```
  -p, --persist-intermediate-table / --np, --no-persist-intermediate-table
                                  Persist the intermediate table to the
                                  filesystem to allow for reuse.  [default: p]
```
```
  --persist-intermediate-table / --no-persist-intermediate-table
                                  Persist the intermediate table to the
                                  filesystem to allow for reuse.  [default:
                                  persist-intermediate-table]
```
